annotation/tests/test_tasks_services.py

```diff
@@ -585,3 +636,1697 @@ def test_filter_tasks_db_no_files_or_jobs(
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=True,
+        task_metrics=[
+            mock_task_metric(1, 2, 0.95),
+            mock_task_metric(1, 3, 0.8),
+            mock_task_metric(2, 3, 0.85),
+        ],
+    )
+    result = compare_agreement_scores(mock_agreement_score_response, min_match)
+    assert result.agreement_score_reached
+
+
+def test_compare_agreement_scores_some_below_min_match(
+    mock_agreement_score_response: Mock,
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.9
+    mock_parse_obj_as.return_value = [MagicMock(spec=ResponseScore)]
+    mock_task_metric.side_effect = (
+        lambda task_from_id, task_to_id, metric_score: MagicMock(
+            spec=TaskMetric,
+            task_from_id=task_from_id,
+            task_to_id=task_to_id,
+            metric_score=metric_score,
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[mock_task_metric(1, 2, 0.95)],
+    )
+    result = compare_agreement_scores(mock_agreement_score_response, min_match)
+    assert not result.agreement_score_reached
+
+
+def test_compare_agreement_scores_empty_response(
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.5
+    mock_parse_obj_as.return_value = []
+    mock_task_metric.return_value = MagicMock(spec=TaskMetric)
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[],
+    )
+    result = compare_agreement_scores([], min_match)
+    assert not result.agreement_score_reached
+    assert result.task_metrics == []
+
+
+def test_save_agreement_metrics(mock_session: Mock):
+    mock_scores = MagicMock()
+    mock_scores.task_metrics = []
+    save_agreement_metrics(mock_session, mock_scores)
+
+    mock_session.bulk_save_objects.assert_called_once_with([])
+    mock_session.commit.assert_called_once()
+
+
+def test_get_annotation_tasks(mock_session: Mock):
+    expected_tasks = [MagicMock(), MagicMock()]
+    mock_query = mock_session.query.return_value.filter.return_value
+    mock_query.all.return_value = expected_tasks
+
+    result = get_annotation_tasks(
+        mock_session, job_id=1, file_id=2, pages=[1, 2, 3]
+    )
+
+    mock_session.query.assert_called_once_with(ManualAnnotationTask)
+    mock_session.query.return_value.filter.assert_called_once()
+    mock_all = mock_session.query.return_value.filter.return_value.all
+    mock_all.assert_called_once()
+    assert result == expected_tasks
```

Comment:
We asserted the returned value which we mocked. So the test didn't test anything except the return functionanlity.

Reply:
When I try to verify arguments in _and(..) stmt: 
def test_get_annotation_tasks(mock_session: Mock):
    expected_tasks = [MagicMock(), MagicMock()]
    mock_query = mock_session.query.return_value.filter.return_value
    mock_query.all.return_value = expected_tasks

    job_id = 1
    file_id = 2
    pages = [1, 2, 3]
    result = services.get_annotation_tasks(
        mock_session, job_id=job_id, file_id=file_id, pages=pages
    )

    mock_session.query.assert_called_once_with(ManualAnnotationTask)
    
    # Construct the expected filter condition correctly
    expected_filter = and_(
        ManualAnnotationTask.job_id == job_id,
        ManualAnnotationTask.is_validation.is_(False),
        ManualAnnotationTask.status == TaskStatusEnumSchema.finished,
        ManualAnnotationTask.file_id == file_id,
        ManualAnnotationTask.pages.contained_by(pages)
    )
    
    # Assert that the filter method was called with the expected conditions
    mock_session.query.return_value.filter.assert_called_once(expected_filter)
    mock_query.all.assert_called_once()
    assert result == expected_tasks
it prints an error: TypeError: and_ expected 2 arguments, got 5